File: server/node/src/controllers/genre.ts
import { Router } from "express";
import { db } from "../app";

const genreRouter = Router();

/**
 * GET genres
 * @description Get all genres
 */
genreRouter.get("/", (req, res) => {
  const genres: string[] = [];
  db.each(
    "SELECT DISTINCT genre FROM song",
    (err, row: { genre: string }) => {
      genres.push(row.genre);
    },
    (err, count) => {
      if (err) console.error(err);

      if (err || !count) res.sendStatus(404);
      res.json(genres);
    }
  );
});

export default genreRouter;

File: server/src/server/controllers/genre.rs
use crate::{get_app_state, types::AppState};
use actix_web::{get, web, HttpResponse, Responder};
use std::sync::{Arc, Mutex};

#[get("")]
pub async fn get_genres(data: web::Data<Arc<Mutex<AppState>>>) -> impl Responder {
    let state = get_app_state!(data);
    let mut genres: Vec<String> = vec![];
    let conn = &state.db;

    match conn.prepare("SELECT DISTINCT genre FROM song") {
        Ok(mut stmt) => {
            if let Ok(iter) = stmt.query_map([], |row| {
                let genre: String = row.get(0)?;
                Ok(genre)
            }) {
                for genre in iter {
                    if let Ok(genre) = genre {
                        genres.push(genre);
                    }
                }
            }
        }
        Err(error) => {
            return HttpResponse::BadRequest().body(format!("{error}"));
        }
    }

    HttpResponse::Ok().json(genres)
}
